packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -824,6 +835,20 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             message.contents = JSON.parse(message.contents);
         }
 
+        // validate client sequence number has no gap. If there is gap, check if there were noops
+        // if there were noops, decrement the gap by number of noops and continue
+        if (this.lastClientSequenceNumber !== undefined && this.lastSubmittedClientId !== undefined && this.lastSubmittedClientId === message.clientId) {
+            const clientSeqNumGap = message.clientSequenceNumber - this.lastClientSequenceNumber - 1;
+            if (clientSeqNumGap > 0) {
+                if (this.noOpCount > 0) {
+                    this.noOpCount -= clientSeqNumGap;
```

Comment:
This looks right for the case where noops are coalesced (i.e. they are dropped by service).
But not all noops are coalesced. If that happens, this.noOpCount will increase when sending, but will never decrease when processing ops here, thus leaving higher value of this.noOpCount forevery, and reducing ability to detect misuse.